ccp/lib/clients/openshift/client.py

```diff
@@ -0,0 +1,344 @@
+            namespace=namespace,
+            bc=bc,
+        )
+        return run_command_exception_on_stderr(cmd=command, shell=False)
+
+    @retry(tries=10, delay=3, backoff=2)
+    def list_build_configs(self, namespace, filter_out=None, selectors=None):
+        """
+        Lists all the BuildConfigs in specified namespace except those
+        requested to be filtered out.
+        :param namespace: The namespace from which to get the build configs.
+        :param filter_out: Default None: A list of BuildConfig names to filter
+        out.
+        :param selectors Filter by selectors, provides list of labels
+        :type namespace: str
+        :type filter_out: list
+        :type selectors: dict
+        :return: A list of BuildConfigs in namespace, after filtering
+        :rtype list
+        :raises Exception
+        :raises subprocess.CalledProcessError
+        :raises ccp.lib.exceptions.CommandOutputError
+        """
+        if not filter_out or not isinstance(filter_out, list):
+            filter_out = []
+        selector_params = "".join(
+            str(" --selector {k}={v}").format(k, v)
+            for k, v in selectors.iteritems()
+        )
+        command = "{base_command} get bc -o name -n {namespace}{s}".format(
+            base_command=self.base_command,
+            namespace=namespace,
+            s=selector_params
+        )
+        out = run_command_exception_on_stderr(cmd=command)
+        if not out.strip():
+            return []
+        out = out.strip().split('\n')
+        return [each for each in out if each not in filter_out]
+
+    def list_builds(self, namespace, status_filter=None, filter_out=None):
+        """
+        Lists builds in a particular namespace, after filtering by status and
+        name.
+        :param namespace: The namespace from which to get the builds
+        :param status_filter: Default None : List of statuses to look for. If
+        provided, only builds with specified statuses are returned.
+        :param filter_out: Default None : List of builds to filter out.
+        :type namespace: str
+        :type status_filter: list
+        :type filter_out: list
+        :return: List of builds that matched all conditions
+        :rtype list
+        :raises Exception
+        :raises subprocess.CalledProcessError
+        :raises ccp.lib.exceptions.CommandOutputError
+        """
+        # Initialize filter_str and close_str to values they will have
+        # if filter_out is not provided.
+        filter_str = ""
+        close_str = "{{end}}"
+
+        # Ensure filter_out is a list
+        if not filter_out:
+            filter_out = []
+
+        if status_filter:
+            conditional = '(ne .status.phase "{}") '
+            condition = ''
+            for phase in status_filter:
+                condition = condition + conditional.format(phase)
+            filter_str = "{{if and %s }}" % condition
+            close_str = "{{end}}{{end}}"
```

Comment:
repeated definitions of `close_str` first occurrence at line 314 in same file

Reply:
Yes. It is meant to be overridden. If filter_out is provided then close_str="{{end}}{{end}}". Otherwise close_str="{{end}}". 

That is why it is being set first to one occurrence of {{end}}, and if filter_out is provided then it is reinitialised to 2 occurrences.